README.md
<p align="center">
    <a href="https://github.com/cjtterabytesoft/toolbar" target="_blank">
        <img src="https://farm1.staticflickr.com/887/27875183957_69a3645a56_q.jpg" height="100px;">
    </a>
    <h1 align="center">Widget ToolBar Gridview. [Yii 3.0/Bootstrap4]</h1>
    <br>
</p>

[![Yii2](https://img.shields.io/badge/Powered_by-Yii_Framework-green.svg?style=flat)](https://www.yiiframework.com/)
[![Build Status](https://scrutinizer-ci.com/g/cjtterabytesoft/toolbar/badges/build.png?b=master)](https://scrutinizer-ci.com/g/cjtterabytesoft/toolbar/build-status/master)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/cjtterabytesoft/toolbar/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/cjtterabytesoft/toolbar/?branch=master)
[![Code Intelligence Status](https://scrutinizer-ci.com/g/cjtterabytesoft/toolbar/badges/code-intelligence.svg?b=master)](https://scrutinizer-ci.com/code-intelligence)
[![Maintainability](https://api.codeclimate.com/v1/badges/074b60f90f8594dc1bb8/maintainability)](https://codeclimate.com/github/cjtterabytesoft/toolbar/maintainability)
[![Total Downloads](https://poser.pugx.org/cjtterabytesoft/toolbar/downloads)](https://packagist.org/packages/cjtterabytesoft/toolbar)


DIRECTORY STRUCTURE
-------------------

```
src/                contains widgets custom application
tests/              contains test widgets
```

REQUIREMENTS
------------

The minimum requirement by this extension that your Web server supports PHP 7.1.

INSTALLATION
------------

If you do not have [Composer](http://getcomposer.org/), you may install it by following the instructions
at [getcomposer.org](http://getcomposer.org/doc/00-intro.md#installation-nix).

[Installation Instructions.](docs/getting-started.md) 

CONFIGURATION
-------------

### WEB SERVER SUPPORT:

- Apache.
- Nginx.
- OpenLiteSpeed.

DOCUMENTATION STYLE GUIDE
-------------------------

This extension follows the documentation described in [Yii Documentation Style Guide](https://github.com/yiisoft/yii2/blob/master/docs/documentation_style_guide.md).


### HOW TO USE RULES CODING STANDARD WITH VSCODE:

- Install squizlabs/php_codesniffer, friendsofphp/php-cs-fixer global:

```
composer global require "squizlabs/php_codesniffer >=2.3.1 <3.0"
composer global require friendsofphp/php-cs-fixer
```

- Copy directory /vendor/yiisoft/yii2-conding-standards/Yii2 to directory path composer global example windows C:\Users\user\AppData\Roaming\Composer\vendor\squizlabs\php_codesniffer\CodeSniffer\Standards

- Add config Vscode:

```
{
    "editor.detectIndentation": false,
    "files.eol": "\n",
    "phpcs.standard": "Yii2",
    "php-cs-fixer.executablePath": "php-cs-fixer",
    "php-cs-fixer.executablePathWindows": "php-cs-fixer.bat",
    "php-cs-fixer.config": ".php_cs;.php_cs.dist",
    "php-cs-fixer.allowRisky": true,
}
```

**LICENSE:**

[![License](https://poser.pugx.org/cjtterabytesoft/toolbar/license)](https://packagist.org/packages/cjtterabytesoft/toolbar)

changelog.md
Widget ToolBar Gridview. [Yii 3.0/Bootstrap4]
=============================================

1.0.0 under development
-----------------------

- Enh 1: Add widget ToolBar.php (tecnoterabyte)

src/ToolBar.php
<?php

/**
 * (c) CJT TERABYTE INC
 * For the full copyright and license information, please view the LICENSE.md
 * file that was distributed with this source code.
 *
 *        @link: https://gitlab.com/cjtterabytesoft/tadweb
 *      @author: Wilmer Arámbula <terabytefrelance@gmail.com>
 *   @copyright: (c) CJT TERABYTE INC
 *     @widgets: [ToolBar]
 *       @since: 1.0
 *         @yii: 3.0
 **/

namespace cjtterabytesoft\widgets;

use cjtterabytesoft\widgets\assets\ToolBarAsset;
use yii\base\Widget;
use yii\bootstrap4\ButtonDropdown;
use yii\helpers\ArrayHelper;
use yii\helpers\Html;
use yii\helpers\Url;

class ToolBar extends Widget
{
	/**
	 * @var string the tag to use to render panel title
	 */
	public $_tag_container_panel_header = 'div';

	/**
	 * @var array the options to use to render panel title
	 */
	public $_options_container_panel_header = ['class' => 'peers bg-primary text-white align-content-center p-15'];

	/**
	 * @var string the tag to use to render panel title left
	 */
	public $_tag_left_panel_header = 'div';

	/**
	 * @var array the options to use to render panel title left
	 */
	public $_options_left_panel_header = ['class' => 'float-left'];

	/**
	 * @var string the tag to use to render panel title rigth
	 */
	public $_tag_rigth_panel_header = 'div';

	/**
	 * @var array the options to use to render panel title rigth
	 */
	public $_options_rigth_panel_header = ['class' => 'float-right ml-auto'];

	/**
	 * @var string use to panel header title
	 */
	public $_panel_header_title = '';

	/**
	 * @var string the tag to use to render panel title icon
	 */
	public $_tag_icon_panel = 'i';

	/**
	 * @var string the label to use to render panel title icon
	 */
	public $_label_icon_panel = '';

	/**
	 * @var array the options to use to render panel title icons
	 */
	public $_options_icon_panel = ['class' => 'fas fa-th'];

	/**
	 * @var string the label panel title
	 */
	public $_title_panel = '';

	/**
	 * @var bool Show/Hidden Summary Label
	 */
	public $_summary = false;

	/**
	 * @var string the tag to use to render container panel button
	 */
	public $_tag_container_panel_button = 'div';

	/**
	 * @var array the options to use to render container panel button
	 */
	public $_options_container_panel_button = ['class' => 'peers bd p-15'];

	/**
	 * @var string the tag to use to render panel button left
	 */
	public $_tag_left_panel_button = 'div';

	/**
	 * @var array the options to use to render panel button left
	 */
	public $_options_left_panel_button = ['class' => 'peers float-left'];

	/**
	 * @var string the tag to use to render panel button rigth
	 */
	public $_tag_rigth_panel_button = 'div';

	/**
	 * @var array the options to use to render panel button rigth
	 */
	public $_options_rigth_panel_button = ['class' => 'peers float-right ml-auto'];

	/**
	 * @var array Templates
	 */
	public $_templates = [];

	/**
	 * @var array Buttons
	 */
	public $_toolbar = [];

	/**
	 * @var bool show/hidden panel button pages
	 */
	public $_button_pages = false;

	/**
	 * @var bool show/hidden panel button export
	 */
	public $_button_export = false;

	/**
	 * @var bool whether the label should be HTML-encoded.
	 */
	public $_encodeLabel = false;

	/**
	 * @var bool whether the JS file should be registered.
	 */
	public $_registerJs = true;

	/**
	 * @var array Options Filter.
	 */
	public $_optionsFilter = [];

	/**
	 * @var string Message Filter Error.
	 */
	public $_messageFilter = '';

	/**
	 * Initializes the widget.
	 * If you override this method, make sure you call the parent implementation first.
	 */
	public function init()
	{
		parent::init();

		if (empty($this->_optionsFilter['message_filter'])) {
			$this->_messageFilter = \yii::t('toolbar', 'Please select one or more items from the list.');
		}

		$iconpanel = $this->renderIcon();
		$titlepanel = $this->renderTitlePanel();

		$this->_panel_header_title = $iconpanel . '&nbsp' . '<b>' . $titlepanel . '</b>';
	}

	/**
	 * Run the widget.
	 * If you override this method, make sure you call the parent implementation first.
	 */
	public function run()
	{
		if (!empty($this->_optionsFilter['id_grid']) && !empty($this->_optionsFilter['id_button']) && !empty($this->_optionsFilter['route_filter'])) {
			$id_grid = $this->_optionsFilter['id_grid'];
			$id_button = $this->_optionsFilter['id_button'];
			$route_filter = $this->_optionsFilter['route_filter'];
			$this->view->registerJs("filterGridSelect('#$id_grid', '#$id_button', '$route_filter', '$this->_messageFilter');");
		}

		if ($this->_registerJs) {
			$toolbar = new ToolBarAsset();
			$toolbar->register($this->view);
		}

		echo $this->renderPanelHeader() . $this->renderPanelBar();
	}

	private function renderButtonExport()
	{
		$button_export = '';
		if ($this->_button_export) {
			$button_export = ButtonDropdown::widget([
				'buttonOptions' => ['class' => 'btn-primary ai-c mL-2'],
				'encodeLabel' => $this->_encodeLabel,
				'label' => Html::tag('i', '', ['class' => 'fas fa-file-export']),
				'options' => ['class' => 'dropleft'],
				'dropdown' => [
					'encodeLabels' => $this->_encodeLabel,
					'items' => [
						Html::tag('h6', \yii::t('toolbar', 'Export Menu'), ['class' => 'dropdown-header']),
						Html::tag('div', '', ['class' => 'dropdown-divider']),
						[
							'label' => Html::tag('i', '', ['class' => 'fas fa-file']) . ' ' . 'Csv',
							'url' => Url::current(['export']),
							'linkOptions' => ['data' => ['method' => 'post', 'params' => ['actionExport' => 'Csv']]],
						],
						[
							'label' => Html::tag('i', '', ['class' => 'fas fa-file-excel']) . ' ' . 'Excel',
							'url' => Url::current(['export']),
							'linkOptions' => ['data' => ['method' => 'post', 'params' => ['actionExport' => 'Excel']]],
						],
						[
							'label' => Html::tag('i', '', ['class' => 'fas fa-file-code']) . ' ' . 'Html',
							'url' => Url::current(['export']),
							'linkOptions' => ['data' => ['method' => 'post', 'params' => ['actionExport' => 'Html']]],
						],
						[
							'label' => Html::tag('i', '', ['class' => 'fas fa-file-pdf']) . ' ' . 'Pdf',
							'url' => Url::current(['export']),
							'linkOptions' => ['data' => ['method' => 'post', 'params' => ['actionExport' => 'Pdf']]],
						],
						[
							'label' => Html::tag('i', '', ['class' => 'fas fa-file-word']) . ' ' . 'Word',
							'url' => Url::current(['export']),
							'linkOptions' => ['data' => ['method' => 'post', 'params' => ['actionExport' => 'Word']]],
						],
					],
				],
			]);
		}

		return $button_export;
	}

	private function renderButtonPages()
	{
		$button_pages = '';

		if ($this->_button_pages) {
			$button_pages = ButtonDropdown::widget([
				'buttonOptions' => ['class' => 'btn-primary ai-c mL-2'],
				'encodeLabel' => $this->_encodeLabel,
				'label' => Html::tag('i', '', ['class' => 'fas fa-cog']),
				'options' => ['class' => 'float-right'],
				'dropdown' => [
					'items' => [
						Html::tag('h6', \yii::t('toolbar', 'Page Size Menu'), ['class' => 'dropdown-header']),
						Html::tag('div', '', ['class' => 'dropdown-divider']),
						['label' => '1', 'url' => Url::current(['index', 'page' => 1, 'per-page' => '1'])],
						['label' => '5', 'url' => Url::current(['index', 'page' => 1, 'per-page' => '5'])],
						['label' => '10', 'url' => Url::current(['index', 'page' => 1, 'per-page' => '10'])],
						['label' => '20', 'url' => Url::current(['index', 'page' => 1, 'per-page' => '20'])],
						['label' => '25', 'url' => Url::current(['index', 'page' => 1, 'per-page' => '25'])],
						['label' => '50', 'url' => Url::current(['index', 'page' => 1, 'per-page' => '50'])],
					],
				],
			]);
		}

		return $button_pages;
	}

	private function renderIcon()
	{
		return Html::tag($this->_tag_icon_panel, $this->_label_icon_panel, $this->_options_icon_panel);
	}

	private function renderPanelBar()
	{
		$panel_button = Html::begintag($this->_tag_container_panel_button, $this->_options_container_panel_button) .
							$this->renderToolBar() .
						Html::endTag($this->_tag_container_panel_button);
		return $panel_button;
	}

	private function renderPanelHeader()
	{
		$pageSize = \yii::$app->params['defaultPageSize'];

		if (\yii::$app->session->get('per-page') !== null) {
			$pageSize = \yii::$app->session->get('per-page');
		}

		$summary = ($this->_summary) ? '{summary}' . ' ' . 'Records per pages: ' . '<b>' . $pageSize . '</b>' : ' ';

		$panel_header = Html::begintag($this->_tag_container_panel_header, $this->_options_container_panel_header) .
							Html::begintag($this->_tag_left_panel_header, $this->_options_left_panel_header) .
								$this->_panel_header_title .
							Html::endTag($this->_tag_left_panel_header) .
							Html::begintag($this->_tag_rigth_panel_header, $this->_options_rigth_panel_header) .
								$summary .
							Html::endTag($this->_tag_rigth_panel_header) .
						Html::endTag($this->_tag_container_panel_header);
		return $panel_header;
	}

	private function renderTitlePanel()
	{
		if (empty($this->_title_panel)) {
			$this->_title_panel = \yii::t('toolbar', 'Gridview ToolBar');
		}

		return $this->_title_panel;
	}

	private function renderToolBar()
	{
		$buttons_left = '';
		$buttons_rigth = '';

		ArrayHelper::setValue($this->_toolbar, 'pages.0', $this->renderButtonPages());
		ArrayHelper::setValue($this->_toolbar, 'export.0', $this->renderButtonExport());

		foreach ($this->_templates as $items => $buttons) {
			foreach ($buttons as $item => $button) {
				if (!empty(ArrayHelper::getValue($this->_toolbar, $button, ''))) {
					switch ($items) {
						case 'left':
							$buttons_left .= implode(',', ArrayHelper::getValue($this->_toolbar, $button, ''));
							break;
						case 'rigth':
							$buttons_rigth .= implode(',', ArrayHelper::getValue($this->_toolbar, $button, ''));
							break;
					}
				}
			}
		}

		return Html::tag($this->_tag_left_panel_button, $buttons_left, $this->_options_left_panel_button) .
			   Html::tag($this->_tag_rigth_panel_button, $buttons_rigth, $this->_options_rigth_panel_button);
	}
}


src/assets/js/toolbar.js
/**
 * (c) CJT TERABYTE INC
 * For the full copyright and license information, please view the LICENSE.md
 * file that was distributed with this source code.
 *
 *        @link: https://github.com/cjtterabytesoft/adminator
 *      @author: Wilmer Arámbula <terabytefrelance@gmail.com>
 *   @copyright: (c) CJT TERABYTE INC
 *          @js: [toolbar]
 *       @since: 1.0
 *         @yii: 3.0
 **/

/***********************************************************************************************************************
 * Filter GridView Select Rows                                                                                         *
 ***********************************************************************************************************************/

(function () {
	filterGridSelect = function(grid, button, route, message) {
		$(button).click(function() {
    		var ids = $(grid).yiiGridView('getSelectedRows');
        
			if (!ids.length) {
				bootbox.alert(message);
				} else if (route) {
					var form = $('<form action=' + route + ' method=\"post\"></form>'),                
						csrfParam = $('meta[name=csrf-param]').prop('content'),
						csrfToken = $('meta[name=csrf-token]').prop('content');

				if (csrfParam) {
					form.append('<input type=\"hidden\" name=' + csrfParam + ' value=' + csrfToken + ' />');
				}
							
				$.each(ids, function(index, id) {
					form.append('<input type=\"hidden\" name=\"' + 'ids' + '[]\" value=' + id + ' />');
				});
        
				form.appendTo('body').submit();
			}
		});
	};
})();

src/Export.php
<?php

/**
 * (c) CJT TERABYTE INC
 * For the full copyright and license information, please view the LICENSE.md
 * file that was distributed with this source code.
 *
 *        @link: https://github.com/cjtterabytesoft/alert
 *      @author: Wilmer Arámbula <terabytefrelance@gmail.com>
 *   @copyright: (c) CJT TERABYTE INC
 *     @widgets: [Export]
 *       @since: 1.0
 *         @yii: 3.0
 **/

namespace cjtterabytesoft\widgets;

use Dompdf\Dompdf;
use Dompdf\Options;
use PhpOffice\PhpSpreadsheet\Spreadsheet;
use PhpOffice\PhpSpreadsheet\Style\Alignment;
use PhpOffice\PhpSpreadsheet\Writer\Xlsx;
use PhpOffice\PhpWord\IOFactory;
use PhpOffice\PhpWord\PhpWord;
use yii\base\InvalidConfigException;
use yii\base\Widget;

/**
 * Export File diferents format Csv, Excel, Html, Pdf, Word.
 */
class Export extends Widget
{
	/**
	 * @var string Charset export csv.
	 */
	public $_csvCharset = 'UTF-8';

	/**
	 * @var \yii\data\DataProviderInterface the data provider for the view. This property is required.
	 */
	public $_dataProvider;

	/**
	 * @var \app\model\searchModel.
	 */
	public $_searchModel;

	/**
	 * @var string tableName.
	 */
	public $_tableName= '';

	/**
	 * @var string Title Export.
	 */
	public $_title = '';

	/**
	 * Initializes the view.
	 */
	public function init()
	{
		parent::init();

		if (empty($this->_dataProvider)) {
			throw new InvalidConfigException('The "dataProvider" property must be set.');
		}

		if (empty($this->_searchModel)) {
			throw new InvalidConfigException('The "dataProvider" property must be set.');
		}

		$this->_tableName = $this->_searchModel->tableName();

		if (empty($this->_title)) {
			$this->_title = $this->_tableName;
		}
	}

	/**
	 * @return file Csv Export.
	 */
	public function exportCsv()
	{
		$csvCharset = $this->_csvCharset;
		$fields = $this->getFieldsKeys($this->_searchModel->exportFields());
		header('Pragma: public');
		header('Expires: 0');
		header('Cache-Control: must-revalidate, post-check=0, pre-check=0');
		header('Content-Description: File Transfer');
		header('Content-Type: text/csv');
		$filename = $this->_tableName . '.csv';
		header('Content-Disposition: attachment;filename=' . $filename);
		header('Content-Transfer-Encoding: binary');
		$fp = fopen('php://output', 'w');
		fwrite($fp, $bom = (chr(0xEF) . chr(0xBB) . chr(0xBF)));
		if ($fp) {
			$items = [];
			$i = 0;
			foreach ($fields as $one) {
				$items[$i] = $one;
				$i++;
			}
			fwrite($fp, implode($items, ',') . "\n");
			$items = [];
			$i = 0;
			foreach ($this->_dataProvider->getModels() as $model) {
				foreach ($this->_searchModel->exportFields() as $one) {
					$item = str_replace('"', '\"', is_string($one) ? $model[$one] : $one($model));
					$items[$i] = ($item) ? '"' . $item . '"' : $item;
					$i++;
				}
				fwrite($fp, implode($items, ',') . "\n");
				$items = [];
				$i = 0;
			}
		}
		fclose($fp);
	}

	/**
	 * @return file Excel Export.
	 */
	public function exportExcel()
	{
		$fields = $this->getFieldsKeys($this->_searchModel->exportFields());
		$objPHPExcel = new Spreadsheet();
		$objPHPExcel->setActiveSheetIndex(0);
		$objPHPExcel->getActiveSheet()->setTitle($this->_title);
		$letter = 65;
		foreach ($fields as $one) {
			$objPHPExcel->getActiveSheet()->getColumnDimension(chr($letter))->setAutoSize(true);
			$objPHPExcel->getActiveSheet()->setCellValue(chr($letter) . '1', $this->_searchModel->getAttributeLabel($one));
			$objPHPExcel->getActiveSheet()->getStyle(chr($letter) . '1')->getAlignment()->setHorizontal(Alignment::HORIZONTAL_CENTER);
			$letter++;
		}
		$row = 2;
		$letter = 65;
		foreach ($this->_dataProvider->getModels() as $model) {
			foreach ($this->_searchModel->exportFields() as $one) {
				$objPHPExcel->getActiveSheet()->setCellValue(chr($letter) . $row, (is_string($one)) ? $model[$one] : $one($model));
				$objPHPExcel->getActiveSheet()->getStyle(chr($letter) . $row)->getAlignment()->setHorizontal(Alignment::HORIZONTAL_LEFT);
				$letter++;
			}
			$letter = 65;
			$row++;
		}
		header('Content-Type: application/vnd.ms-excel');
		$filename = $this->_tableName . '.xlsx';
		header('Content-Disposition: attachment;filename=' . $filename);
		header('Cache-Control: max-age=0');
		$objWriter = new Xlsx($objPHPExcel);
		$objWriter->save('php://output');
	}

	/**
	 * @return file Htlm Export.
	 */
	public function exportHtml()
	{
		$fields = $this->getFieldsKeys($this->_searchModel->exportFields());
		$phpWord = new PhpWord();
		$section = $phpWord->addSection();
		$section->addTitle($this->_title);
		$table = $section->addTable(['name' => 'Tahoma', 'size' => 10, 'align' => 'center',]);
		$table->addRow(300, ['exactHeight' => true]);
		foreach ($fields as $one) {
			$table->addCell(
				1500,
				['bgColor' => 'eeeeee', 'borderTopSize' => 5, 'borderRightSize' => 5, 'borderBottomSize' => 5,'borderLeftSize' => 5, 'valign' => 'center']
			)->addText($this->_searchModel->getAttributeLabel($one), ['bold' => true, 'size' => 10], ['align' => 'center']);
		}
		foreach ($this->_dataProvider->getModels() as $model) {
			$table->addRow(300, ['exactHeight' => true]);
			foreach ($this->_searchModel->exportFields() as $one) {
				$table->addCell(
					1500,
					['borderTopSize' => 1, 'borderRightSize' => 1, 'borderBottomSize' => 1, 'borderLeftSize' => 1, 'valign' => 'center']
				)->addText('<p style="margin-left: 10px;">' . (is_string($one)) ? $model[$one] : $one($model) . '</p>', ['bold' => false, 'size' => 10], ['align' => 'right']);
			}
		}
		header('Content-Type: application/html');
		$filename = $this->_tableName . '.html';
		header('Content-Disposition: attachment;filename=' . $filename . ' ');
		header('Cache-Control: max-age=0');
		$objWriter = IOFactory::createWriter($phpWord, 'HTML');
		$objWriter->save('php://output');
	}

	/**
	 * @return file Pdf Export.
	 */
	public function exportPdf()
	{
		$fields = $this->getFieldsKeys($this->_searchModel->exportFields());
		$options = new Options();
		$options->set('defaultFont', 'times');
		$dompdf = new Dompdf($options);
		$html = '<html><body>';
		$html .= '<h1>' . '<center>' . $this->_title . '</center>' . '</h1>';
		$html .= '<table width="100%" cellspacing="0" cellpadding="0">';
		$html .= '<tr style="background-color: #ececec;">';
		foreach ($fields as $one) {
			$html .= '<td style="border: 2px solid #cccccc; text-align: center; font-weight: 500;">' . $this->_searchModel->getAttributeLabel($one) . '</td>';
		}
		$html .= '</tr>';
		foreach ($this->_dataProvider->getModels() as $model) {
			$html .= '<tr>';
			foreach ($this->_searchModel->exportFields() as $one) {
				switch (true) {
					case is_string($one):
						$html .= '<td style="border: 1px solid #cccccc; text-align: left; font-weight: 300; padding-left: 10px;">' . $model[$one] . '</td>';
						break;
					default:
						$html .= '<td style="border: 1px solid #cccccc; text-align: left; font-weight: 300; padding-left: 10px;">' . $one($model) . '</td>';
						break;
				}
			}
			$html .= '</tr>';
		}
		$html .= '</table>';
		$html .= '</body></html>';
		$dompdf->loadHtml($html);
		$dompdf->setPaper('letter', 'landscape');
		$dompdf->render();
		$dompdf->stream($this->_tableName . '_' . time());
	}

	/**
	 * @return file Word Export.
	 */
	public function exportWord()
	{
		$fields = $this->getFieldsKeys($this->_searchModel->exportFields());
		$phpWord = new PhpWord();
		$phpWord->getCompatibility()->setOoxmlVersion(15);
		$section = $phpWord->addSection();
		$sectionStyle = $section->getSettings();
		$sectionStyle->setLandscape();
		$sectionStyle->setBorderTopColor('C0C0C0');
		$sectionStyle->setMarginTop(300);
		$sectionStyle->setMarginRight(300);
		$sectionStyle->setMarginBottom(300);
		$sectionStyle->setMarginLeft(300);
		$phpWord->addTitleStyle(1, ['name' => 'HelveticaNeueLT Std Med', 'size' => 16], ['align' => 'center']);
		$section->addTitle($this->_title);
		$table = $section->addTable(
			['name' => 'Tahoma', 'align' => 'center', 'cellMarginTop' => 30, 'cellMarginRight' => 30, 'cellMarginBottom' => 30,	'cellMarginLeft' => 30]
		);
		$table->addRow(300, ['exactHeight' => true]);
		foreach ($fields as $one) {
			$table->addCell(
				1500,
				['bgColor' => 'eeeeee', 'borderTopSize' => 5, 'borderRightSize' => 5, 'borderBottomSize' => 5,'borderLeftSize' => 5, 'valign' => 'center']
			)->addText($this->_searchModel->getAttributeLabel($one), ['bold' => true, 'size' => 10], ['align' => 'center']);
		}
		foreach ($this->_dataProvider->getModels() as $model) {
			$table->addRow(300, ['exactHeight' => true]);
			foreach ($this->_searchModel->exportFields() as $one) {
				$table->addCell(
					1500,
					['borderTopSize' => 1, 'borderRightSize' => 1, 'borderBottomSize' => 1, 'borderLeftSize' => 1, 'valign' => 'center']
				)->addText(is_string($one) ? $model[$one] : $one($model), ['bold' => false, 'size' => 10], ['align' => 'left']);
			}
		}
		header('Content-Type: application/vnd.ms-word');
		$filename = $this->_tableName . '.docx';
		header('Content-Disposition: attachment;filename=' . $filename . ' ');
		header('Cache-Control: max-age=0');
		$objWriter = IOFactory::createWriter($phpWord, 'Word2007');
		$objWriter->save('php://output');
	}

	/**
	 * @return array Fields List.
	 */
	private function getFieldsKeys($fieldsSended)
	{
		$fields = [];
		$i = 0;

		foreach ($fieldsSended as $key => $value) {
			switch (true) {
				case is_int($key):
					$fields[$i] = $value;
					break;
				default:
					$fields[$i] = $key;
					break;
			}
			$i++;
		}

		return $fields;
	}
}
